### PR TITLE
[JFG-948] Master to JaaS communication

### PR DESCRIPTION
**Pull request by a-badaev**

### CODE REVIEW

File: jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java [hunk 1]
```diff
@@ -82,33 +75,33 @@ public TestEnvironmentRestController(TestEnvironmentService testEnvironmentServi
...
+            + "present and valid. To obtain this cookie POST to /envs must be performed firstly. "
+            + "This cookie is valid only for Test Environment with "
             + "envId which was specified in POST request body.")
     @ApiResponses(value = {
             @ApiResponse(code = 202, message = "Update is successful."),
             @ApiResponse(code = 404, message = "Test Environment with provided envId or sessionId is not found."),
-            @ApiResponse(code = 400, message = "Environment-Session cookie is not present.")})
-    public ResponseEntity<?> updateTestEnvironment(@PathVariable String envId, @RequestBody TestEnvironmentEntity testEnv,
-                                                   HttpServletRequest request, HttpServletResponse response) {
+            @ApiResponse(code = 404, message = TestEnvUtils.SESSION_COOKIE + " cookie is not present.")})
```
Comment (dlatnikov) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java]: have we finally agreed on this: 400 -> 404?

File: jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java [hunk 2]
```diff
@@ -132,35 +125,23 @@ private void validateTestEnv(TestEnvironmentEntity testEnv) {
...
-                if (cookie.getName().equals(ENVIRONMENT_SESSION_COOKIE)) {
-                    return cookie.getValue();
-                }
-            }
-        }
-        return null;
+    private Cookie getSessionCookie(TestEnvironmentEntity testEnv) {
+        Cookie cookie = new Cookie(TestEnvUtils.SESSION_COOKIE, testEnv.getSessionId());
+        cookie.setMaxAge(environmentsTtlMinutes * 60);
+
```
Comment (anton-antonenko) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java]: Expires time won't be updated without `cookie.setPath("/");`
Check it, please, and fix if I'm right.

File: jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java [hunk 3]
```diff
@@ -82,33 +75,33 @@ public TestEnvironmentRestController(TestEnvironmentService testEnvironmentServi
...
-        String sessionId = getSessionCookie(request);
-        if (sessionId == null)
-            throw new TestEnvironmentNoSessionException();
         if (!testEnvService.existsWithSessionId(envId, sessionId))
             throw new TestEnvironmentSessionNotFoundException(envId, sessionId);
 
         testEnv.setEnvironmentId(envId);
-        TestEnvironmentEntity updated = testEnvService.update(testEnv);
-        setExpiresHeader(response, updated);
-        setSessionCookie(response, updated, environmentsTtlMinutes * 60);
```
Comment (anton-antonenko) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java]: Why did you remove it? Cookie Expires won't be updated and it'll expire very soon.

File: jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java [hunk 4]
```diff
@@ -82,33 +75,33 @@ public TestEnvironmentRestController(TestEnvironmentService testEnvironmentServi
...
-    public ResponseEntity<?> updateTestEnvironment(@PathVariable String envId, @RequestBody TestEnvironmentEntity testEnv,
-                                                   HttpServletRequest request, HttpServletResponse response) {
+            @ApiResponse(code = 404, message = TestEnvUtils.SESSION_COOKIE + " cookie is not present.")})
+    public ResponseEntity<?> updateTestEnvironment(@CookieValue(TestEnvUtils.SESSION_COOKIE) String sessionId,
+                                                   @PathVariable String envId,
+                                                   @RequestBody TestEnvironmentEntity testEnv) {
         if (!testEnvService.exists(envId))
             throw ResourceNotFoundException.getTestEnvResourceNfe();
 
-        String sessionId = getSessionCookie(request);
```
Comment (anton-antonenko) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java]: Why did you remove this check?

File: jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java (hunk 1 above)
Comment (anton-antonenko) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java]: You've removed this check, so this message is not valid. 
Also, you have 2 different responses with same code - are you sure that it will be displayed correctly in Swagger?

File: chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        boolean posted = false;
+        do {
+            try {
+                responseEntity = restTemplate.postForEntity(envsUri, testEnvironmentEntity, String.class);
+                posted = true;
+            } catch (HttpClientErrorException e) {
+                if (!isEnvIdUnacceptable(e)) {
+                    throw e;
+                }
+                envId = UUID.randomUUID().toString();
```
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: One of the reasons we got HttpClientErrorException  here - ID doesn't match to the pattern.

File: chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+                testEnvironmentEntity.setEnvironmentId(envId);
```
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: done

File: chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+            setName(this.getClass().getName());
+        }
+        
+        private RequestEntity<TestEnvironmentEntity> buildPendingRequestEntity() {
+            TestEnvironmentEntity testEnvEntity = buildTestEnvironmentEntityWith(TestEnvironmentStatus.PENDING);
+            HttpHeaders httpHeaders = new HttpHeaders();
+            httpHeaders.add(HttpHeaders.COOKIE, sessionCookie);
+            URI updatesUri = UriComponentsBuilder.newInstance()
+                                                 .uri(jaasEndpoint)
+                                                 .path("/envs")
```
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: you can't give meaningful name to such a constant shorter then "/envs" so it will be more verbose

File: jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java (hunk 1 above)
Comment (a-badaev) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java]: This check is still there but now it is done by Spring and it will throw 400 as well. So, gonna fix

File: jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java (hunk 4 above)
Comment (a-badaev) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/rest/TestEnvironmentRestController.java]: it can't be null now, can it?

File: chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        private ResponseEntity<String> doUpdateStatus() throws InterruptedException {
+            int retries = 0;
+            HttpServerErrorException exception;
+            do {
+                try {
+                    return restTemplate.exchange(requestEntity, String.class);
+                } catch (HttpServerErrorException e) {
+                    exception = e;
+                }
+                Thread.sleep(1000 * 60); // Sleep 1 minute before retry
```
Comment (anton-antonenko) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: TimeUnit.MINUTES.sleep(1); could be better here.

Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: ok. I like it

File: chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+                    LOGGER.error("Finishing work due to HTTP client exception.", e);
+                    standBy = false;
+                }
+            }
+        }
+        
+        private ResponseEntity<String> doUpdateStatus() throws InterruptedException {
+            int retries = 0;
+            HttpServerErrorException exception;
+            do {
```
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: it is a retry loop

File: chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java [hunk 10]
```diff
@@ -155,24 +159,45 @@ private static void launchMaster(final URL directory) {
...
+                    if (isStandByMode) {
+                        log.info("Starting Master in stand by mode...");
+            
+                        MasterToJaasCoordinator masterToJaasCoordinator = new MasterToJaasCoordinator(
+                                environmentProperties.getProperty("realtime.environment.id"),
+                                environmentProperties.getProperty("realtime.jaas.endpoint"), Integer.parseInt(
+                                environmentProperties.getProperty("realtime.status.report.interval.seconds")),
+                                getAvailableConfigurations(directory)
+                        );
+                        masterToJaasCoordinator.register();
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java]: please, add INFO log that we have registered to jaas at endpoint
